api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto

```diff
@@ -22,5 +26,37 @@ message XffConfig {
   // determining the origin client's IP address. The default is zero if this option
   // is not specified. See the documentation for
   // :ref:`config_http_conn_man_headers_x-forwarded-for` for more information.
+  //
+  // If ``xff_trusted_cidrs`` is configured, this field is not used.
   uint32 xff_num_trusted_hops = 1;
+
+  // The `CIDR <https://tools.ietf.org/html/rfc4632>`_ ranges to trust when
+  // evaluating the remote IP address to determine the origin client's IP address.
+  // This is used instead of
+  // :ref:`use_remote_address <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.use_remote_address>`.
+  // When the remote IP address matches a trusted CIDR and the
+  // :ref:`config_http_conn_man_headers_x-forwarded-for` header was sent, the last
+  // IP address from ``x-forwarded-for`` is used as the original client address.
+  // If no ``x-forwarded-for`` header was sent, the remote IP is considered the real IP.
+  //
+  // If ``recurse`` is set the :ref:`config_http_conn_man_headers_x-forwarded-for`
+  // header is also evaluated against the trusted CIDR list, and the last non-trusted
+  // address is used as the original client address.
+  //
+  // This is typically used when requests are proxied by a
+  // `CDN <https://en.wikipedia.org/wiki/Content_delivery_network>`_.
+  //
+  // If this field is configured, ``xff_num_trusted_hops`` is not used.
+  XffTrustedCidrs xff_trusted_cidrs = 2;
+}
+
+message XffTrustedCidrs {
```

Comment:
I think any updates in this area of API should also be spelled out in docs/root/configuration/http/http_conn_man/headers.rst

that's probably a better place to answer my recurse question too btw :-)

Reply:
I added a small note to those docs mentioning to use the extension if you need to use trusted CIDRs, but I wasn't sure how much "non-core" information should go into those docs when it otherwise talks about configuration that's directly in HCM.

Do you think I should add more than the note, or expand that note?

Reply:
All of this code/config is so confusing, and such a security landmine, that it's quite hard to wrap my head around what this all does and why I would use it.

Can you add significant fleshed out examples either here or better in https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#config-http-conn-man-headers-x-forwarded-for where there are other examples? This would allow us to see various worked examples of what the config does and why I would want to use it, and then we can review the implementation? Thank you.

/wait

Reply:
Yep, absolutely! I've been working on that recently and trying to do some more local testing to make sure I've really got everything right based on Alyssa's feedback.

The TL;DR is to make it work in a similar way to Nginx's [realip module](https://nginx.org/en/docs/http/ngx_http_realip_module.html) as per #21639.

Just to confirm something:

> All of this code/config is so confusing, and such a security landmine, that it's quite hard to wrap my head around what this all does and why I would use it.

Are you saying that my code is confusing, or the existing xff handler, or both? If it's mine I'd love to know how I can make it less confusing. I also posted a (slightly long-winded) question in Slack the other day, if either of you have a chance to take a look at that.

Reply:
No it isn't your code, it's this entire part of the codebase. Any time anyone touches it we go around in circles because it is so confusing and the security implications are so profound. Clear examples and a lot of documentation helps a lot. Thank you!